plutus-benchmark/src/Plutus/Benchmark/Queens.hs

```diff
@@ -0,0 +1,553 @@
+{- nofib/spectral/constraints converted to Plutus.
+   Renamed to avoid conflict with existing package. -}
+
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE NamedFieldPuns        #-}
+{-# LANGUAGE NoImplicitPrelude     #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE ScopedTypeVariables   #-}
+
+{-# OPTIONS_GHC -fwarn-missing-signatures     #-}
+{-# OPTIONS_GHC -fno-warn-unused-imports      #-}
+
+{-# OPTIONS_GHC -fexpose-all-unfoldings       #-}
+{-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
+{-# OPTIONS_GHC -fno-spec-constr              #-}
+{-# OPTIONS_GHC -fno-strictness               #-}
+{-# OPTIONS_GHC -fno-worker-wrapper           #-}
+
+module Plutus.Benchmark.Queens where
+
+{- Andrew Tolmach and Thomas Nordin's contraint solver
+
+	See Proceedings of WAAAPL '99
+-}
+
+import           Control.Monad                (forM_)
+import           Data.Char                    (isSpace)
+import qualified Prelude
+import           System.Environment
+
+import           Language.PlutusCore          (Name (..))
+import qualified Language.PlutusCore.Pretty   as PLC
+import           Language.PlutusCore.Universe
+import qualified Language.PlutusTx            as Tx
+import           Language.PlutusTx.Prelude    as TxPrelude hiding (head, notElem, tail)
+import           Language.UntypedPlutusCore
+
+
+-----------------------------
+-- The main program
+-----------------------------
+
+{- kwxm: This program computes the number of mutually non-attacking arrangements
+   of n queens on an n x n chessboard.  It's quite sophisticated: there's a
+   generic constraint solver that takes an algorithm as a parameter.  There are
+   five different algorithms, and the original behaviour was to try each of
+   these on a board if size NxN, where N is a parameter supplied by the user.
+   The Haskell version can run all five algorithms one after the other hundreds
+   of times on an 8x8 board in a few seconds.
+
+   Compiling into Plutus, I tried to run all five algorithms one after the other
+   (in a sinlge run of the program) on a 5x5 board on the CEK machine on an 8 GB
+   machine.  The program was just able to complete in about 23s, but with some
+   swapping.
+
+   The resource consumption isn't so bad for the individual algorithms.  Here
+   are figures for the time and memory usage for one run of the Plutus version
+   of each algorithm on a 5x5 board on the same machine:
+
+   bt:    2.9s, 1.2 GB
+   bm:    2.5s, 830 MB
+   bjbt:  2.7s, 933 MB
+   bjbt': 3.0s, 1.1 GB
+   fc:    7.5s, 3.2 GB
+
+   I suggest keeping the entire program, but only using a single algorithm for
+   benchmarking for the time being.  If we're investigate the behaviour of
```

Comment:
Are those space problems due to ever-growing environments? We do drop environments _sometimes_, can we somehow compose those five algorithms in a way that makes forces the current environment to be dropped when execution moves from one algorithm to the next one?

Reply:
Good question.  I think we can't do that inside a single PLC program, because after you've run the first algorithm it's not obvious that you're never going to refer to it (or anything in the relevant environment) again.  It certainly works if you use a separate invocation of the machine for each algorithm.